Airside_Functions.cpp

```diff
@@ -0,0 +1,468 @@
+/**************************************************************************************************/
+// Author: Jingting Liu
+// April 9th, 2021
+
+// to run the test without cmake:
+// gcc -g Airside_Functions.cpp
+// a.out
+
+// to run the test with cmake:
+// mkdir build, cd build
+// cmake ..
+// make
+// FUNCTIONS (executing the generated target, which is FUNCTIONS in this case)
+
+// The encoder and decoders are implemented based on Mavlink2 (not Mavlink1)
+// refer to this page for the stucture of mavlink messages 
+// https://mavlink.io/en/guide/serialization.html
+
+// The decoder and encoder only support GPS and gimbal control, other simpler commands will be taken
+// care of by the Xbee communication directly
+/**************************************************************************************************/
+
+#include "Mavlink2_lib/common/mavlink.h"
+#include "Airside_Functions.hpp"
+
+mavlink_decoding_status_t Mavlink_airside_decoder(PIGO_Message_IDs_e* type, uint8_t incomingByte, uint8_t *telemetryData)
+{
+    int channel = MAVLINK_COMM_0; //mavlink default one channel
+    PIGO_Message_IDs_e decoded_message_type = MESSAGE_ID_NONE;
+    mavlink_decoding_status_t decoding_status = MAVLINK_DECODING_INCOMPLETE;
+
+    mavlink_status_t status;
+    memset(&status, 0x00, sizeof(mavlink_status_t));
+
+    mavlink_message_t decoded_msg;
+    memset(&decoded_msg, 0x00, sizeof(mavlink_message_t));
+
+    // this function parses the incoming bytes, and the decoded_msg will get filled when the full message is received
+    // more details about the parser function: http://docs.ros.org/en/indigo/api/mavlink/html/include__v2_80_2mavlink__helpers_8h.html#ad91e8323cefc65965574c09e72365d7d
+    uint8_t message_received = mavlink_parse_char(channel, incomingByte, &decoded_msg, &status);
+
+    if (message_received)
+    {
+        if (telemetryData == NULL)
+        {
+            return MAVLINK_DECODING_FAIL;
+        }
+
+        switch(decoded_msg.msgid)
+        {
+            case MAVLINK_MSG_ID_GLOBAL_POSITION_INT: // ID for GLOBAL_POSITION_INT, 33
```

Comment:
Currently the switch only has one single case, that's cuz I removed the other cases and decided to use the GPS structure for everything. The reason why I didn't remove it is that I figured it would be easy if we want to add other cases in future. Ideally, I'd like to remove it after testing with the ground side decoders, and incorporating this with the Xbee stuff to make sure that using the GPS struct to carry all types of data works for sure

For the if statement thing, yeah I used to use switch for those, but I had a reason to change it, though now I can't seem to remember why... Maybe cuz I make some more changes to the way I assigned the IDs so some previous problem went away... Anyways, yeah I'll change it so I will use switches again. Thanks for pointing it out!